Grid based movement

Exonoesis opened the pull request:
## What is the purpose of these changes?
LDtk third-party dependencies were becoming a problem when it came to keeping all parts of the project up-to-date. Also I wanted to switch to a grid-based movement system to better emulate more traditional RPG games.

## What has changed?
- LDtk library has been removed from the project dependencies
- Custom functions that create tile maps directly from .tmx files, to replace the use of LDtk and its library
- Various bug fixes and project-wide optimizations
- Removed dialogue triggers (will be reworked shortly into a fully fledged dialogue system)

## How can I verify this works?
- Run the game
- Press Play
- See that the map loads as before
- Lightly tap WASD
- See that the character now moves in full individual squares
- Hold WASD
- See the character smoothly traverses as well
- Explore the main map
- See that the camera follows the player and does not go past the edges of the map
- Place the character at the front of the cabin door (between the bucket and workbench)
- Press up to "face" the door and press E
- See that the cabin interior loads
- Traverse here as well; place the character behind the plants, top of the bed, or back edges of the dining room table to see proper layering

------------------------

Closes #47 
Closes #54

Exonoesis committed:
Features Files

- Added feature files, including scenario

Exonoesis committed:
Testability Refactor

- Removed Main.rs in favor of a lib and bin approach
- Made fn placeholders for grid-based-movement tests

Exonoesis committed:
Transition to accessing map directly via json

- Added json reader and tests to load LDtk maps directly, circumventing the use of third-party libraries
- Added small testing map

Exonoesis committed:
Work on spawning map

- Created Tile struct
- Wrote spawn_map function

Exonoesis committed:
Added Grid Test

- Wrote tests to verify the map is in a grid

Exonoesis committed:
Cargo fmt

Exonoesis committed:
Added new test map

- Fixed some issues with LDtk map (even though it's being replaced soon)
- Added new test map that includes tiles for collision

Exonoesis committed:
Changed function and variable names

- Restructured function and variable names to read more smoothly

Exonoesis committed:
Added sprites to tiles

- Created struct to hold a reference to a tiles image
- Added function to assign spritesheet and sprite index to each tile
- Wrote associated tests including maps with one or multiple spritesheets

Exonoesis committed:
Cleaned up repetitive functions

- Used regular expressions to combine similar test conditions
- Added helper function to get sprite sheet filename

divark committed:
Add Wayland dependencies for CI.

divark committed:
Update checkout to v4 for all workflows.
Add Wayland dependencies to rest of workflows.

Exonoesis committed:
Updated function names

- Updated test function names to reflect their generalized nature

Exonoesis committed:
Blank image scenario

- Added test that checks for a tile with no image component

Exonoesis committed:
Multiple Layer Scenario

- Added test that checks if the map has multiple layers and are overlapping correctly

Exonoesis committed:
Adapter Feature

- Added beginnings of functions to create and test structs to act as a bridge for rendering tiles

- Known Issue: Code does not currently run due to a resource sharing issue

Exonoesis committed:
Adapter Feature Fixed

- Added needed plugins for TextureAtlasLayout
- Capture sprite sheet width and height within TileTexture
- Added assertions for sprite sheet height and width within sprite sheet correctness tests
- Fixed some loops that weren't looping the correct amount of times
- Added RenderTileBundle tests

Exonoesis committed:
Code Cleanup

- Added comments
- Removed some nesting
- Renamed get_tiles to get_map_tiles for additional clarity

Exonoesis committed:
Bevy path trimming

- Created function to trim absolute paths to have root at the assets folder (as bevy expects)

Exonoesis committed:
Cleaned up function

- Removed unneeded variable

Exonoesis committed:
Changed path trimming function

- Altered method for bevy path conversion
- Added additional tests for path variations

Exonoesis committed:
Coordinate and dimension abstraction

- Made struct hold three-dimensional spacial coordinates in pixels
- Made struct to hold two-dimensional spacial measurements in pixels

Exonoesis committed:
Testing Refactor

- Made steps in scenarios more detailed to aid re-usability
- Created class structs to ease access of information and make function scope clearer
- Renamed several variables for clarity
- Changed calls to clone() to be as_ref() where possible
- Simplified several tests by adding functions
- Utilized regex in many tests to aid re-usability

Exonoesis committed:
Added Y-axis flip function

- Added PxDimensions to Tilemap
- Renamed Tilemap dimensions to grid_dimensions for clarification
- Added function to get map's pixel dimensions
- Added function to check if RenderTile overlaps with a tile on Tiled map
- Added test to verify sucessful y-axis flip

Exonoesis committed:
Tilemap Render Integration

- Moved non-test related code into map.rs for use in the main code base
- Added getter functions that were missing
- Hooked up new map rendering functionality
- Updated Camera render function to spawn at map center
- Implemented various Clippy recommended code optimizations

Exonoesis committed:
Tile Types

- Added types to tiles to identify special properties
- Types are defined by the name of the layer within Tiled, matching on certain keywords

Exonoesis committed:
Change to 3D cord system

- Renamed MapGridDimensions -> GridDimensions for reusability
- Fixed some inconsistencies where rows were listed before columns
- Created function to convert 3D cords into 1D indices
- Converted all tilemap feature tests to take tile references as 3D coordinates

Exonoesis committed:
Grid based movement begins

- RenderTile is no longer an option, one is always spawned even if tile is empty (it is given an invisible image)
- Spritesheet logic has been pulled out of RenderTile creation logic
- Reworded grid-based-movement feature file and started to write function for first scenario

divark committed:
Working on grid-based movement

- Implemented player moving to the right
- Changed loading map to take in a map name rather than being hard coded
- Changed grid_cords in various locations to be XyzCords for clarity
- RenderTiles now hold a tile type, its grid coordinates, and the dimensions of a generic game tile
- Cleaned up testing code for tilemap

divark committed:
Remaining directions & Fixes

- Added remaining movement directions
- Made helper function to calculate distance
- Started to implement fix for excessive sliding
- Created testing specific LevelsPlugin
- Created function to load a default starting map
- Created a function to despawn the map
- Fixed bug in 3d to 1d index conversion

[KNOWN ISSUES]
- Player currently speeds to target too quickly then bounces violently until the end of the timer

divark committed:
Grid based movement completed

- Finished fix for excessive sliding
- Fine tuned movement speed

Exonoesis committed:
Starting work on collision

- Created collision feature file and test file
- Made new TileType for collision tiles
- Made collision tiles render with visibility off
- Began work on creating a collection of collision tiles

Exonoesis committed:
Completed gathering collision tiles

- Added getters to XyzCords for x, y, and z values
- Added second new constructor to XyzCords so that x and y values can be given as u32 or usize
- Added getter to RenderTile for grid_coordinates
- Created collision_collection HashSet to hold the maps collision tiles

Exonoesis committed:
Collision Benchmarking

- Moved collision detection logic into main map function
- Added Benchmarking tools to project
- Created Benchmarking tests for collision tile lookup times at various map grid sizes

Exonoesis committed:
Player Grid Coords Updating

- Target now has concept of physical and logical position (needed for ease of collision detection)
- Player's grid coordinate now updates after movement occurs

Exonoesis committed:
Implemented Collision

- Hooked up systems to make player honor collision by stopping movement

Exonoesis committed:
Started implementing camera following player and bounding

- Removed LevelDimensions, replacing it with the maps PxDimensions
- Fixed Collision Bug
- Removed Old Code and Tests

Exonoesis committed:
Camera code rewrite

- Split camera moving logic into two parts; follow_player, and bound camera
- Began work on camera test files & functions related to following the player
- Made mock_game file for shared game function access across test files

Exonoesis committed:
White dot fix

Exonoesis committed:
Finished camera code, including bounds

- Added getters for columns and rows for GridDimensions
- Implemented camera following player and staying within level boundaries

Exonoesis committed:
Camera Testing Updates and Fixes

- Added/Updated some camera tests

Exonoesis committed:
Camera Centering Fix

- Fixed camera not being properly centered when map is narrower or shorter than camera (and both, by proxy)

Exonoesis committed:
Player Bounds and Audio Reintegration

- Changed set_logical_destination to return an option type; thus allowing out of bounds destinations to be returned as invalid (none)
- Renamed audio file to play on new testing level
- Made EventReader to get level name in order to load music
- Made EventWriter to broadcast SFX events

Exonoesis committed:
Made interactive markers

- Implemented basic markers for interactive locations
- Created a collection to hold all markers on map

Exonoesis committed:
Interactives Scenarios

- Wrote test scenarios for interactives

Exonoesis committed:
Created containing function for binary search on markers

- Created Proximity enum to report to binary search
- Implemented containing function that determines if a given x,y is within the range of a given marker
- Made helper function to convert string into a proximity

Exonoesis committed:
Binary Search Prep

- Added equality and ordering to XyzCords, PxDimensions, and InteractiveMarker
- Fixed marker function being passed a marker
- Restructured nested if's to no longer be nested
- Removed add function
- Implemented replacement from_markers function
- Created wire frame for get_marker_from_position function

Note: Interactives test will not pass for this push! The option type from get_marker_from_position is currently returning None by default; which is being unwrapped by the tests and will cause a panic

Exonoesis committed:
Binary search for markers complete

- Fixed off by one-error in marker size calculation
- Implemented binary search algorithm, using containing for higher/lower reporting

Exonoesis committed:
Interactives Benchmarking

- Added benchmark testing for get_marker_from_position function
- Actually fixed off by one error in binary search
- Removed flooring function and unneeded floating point conversions

Exonoesis committed:
More marker search tests

- Added more tests
- Regex'd various test functions

Exonoesis committed:
Cabin Interior

- Added Cabin Interior map

Exonoesis committed:
Player Facing Direction and Interactive Type setup

- Added MovementDirection component to player
- Made helper function to convert Transforms into XyzCords
- Setup for adding types to interactive markers

Exonoesis committed:
Update multiple_marker_test.tmx

Exonoesis committed:
Implement Interactive Transitions

- Removed InteractiveType of Text
- Added getters in InteractiveMarker for type name and path/value
- Created function to add interactive type to markers on creation

Exonoesis committed:
Transition Interactions Complete

- Finished implementing level transition code
- Hooked up level transition code to player button press (E)
- Removed LDtk third party dependency

Exonoesis committed:
Updated project example gifs

Exonoesis committed:
Update CHANGELOG.md

divark commented on assets/map/hh_world.ldtk:
We can delete this now, right?

Exonoesis replied:
It still contains my first draft of the map for chapter 0, so I was thinking of keeping it until that map it recreated. However, I can move it to an external location or take pictures. Yes, I can remove this now.

Exonoesis replied:
On further inspection, tests involving Missing File Finder are reliant on the LDtk map. These will need to be reworked before removing the LDtk map

divark commented on src/mechanics/input.rs:
```diff
@@ -1,628 +1,594 @@
+//    }
+//}
+
+//pub fn move_entity(
+//    mut entity_query: Query<
+//        (&mut Transform, &DirectionFacing, &mut MovementIntent),
+//        Changed<MovementIntent>,
+//    >,
+//    tile_query: Query<&EntityInstance>,
+//    level_dimension: Res<LevelDimensions>,
+//    mut entity_movement_broadcast: EventWriter<PlayerMovementActions>,
+//) {
+//    if entity_query.is_empty() {
+//        return;
+//    }
+//
+//    let collision_tiles = tile_query
+//        .iter()
+//        .filter(|&tile| !tile.field_instances.is_empty())
+//        .filter(|&tile| {
```
We should be able to remove this commented out function now, unless we have this here as reference for something else in the future.

Exonoesis replied:
Looks like we've replaced all parts of this function, it should be fine to remove it.

divark commented on src/mechanics/input.rs:
```diff
@@ -1,628 +1,594 @@
-        assert_eq!(expected_transform, actual_transform);
+pub fn interact_entity(
+    input: Res<ButtonInput<KeyCode>>,
+    mut interactive_event_wrier: EventWriter<PlayerInteraction>,
```
```suggestion
    mut interactive_event_writer: EventWriter<PlayerInteraction>,
```

Exonoesis replied:
Oops

divark commented on src/mechanics/input.rs:
```diff
@@ -1,628 +1,594 @@
-
-        assert_eq!(expected_transform, actual_transform);
-    }
+    interactive_event_wrier.send(PlayerInteraction);
```
```suggestion
    interactive_event_writer.send(PlayerInteraction);
```

Exonoesis replied:
Continued oops

divark commented on src/mechanics/input.rs:
```diff
@@ -1,628 +1,594 @@
+    interactive_event_wrier.send(PlayerInteraction);
 }
+
+// pub fn display_interactive_message(mut interactible_event_reader: EventReader<InteractionEvent>) {
+//     for interaction_command in interactible_event_reader.read() {
+//         let command = &interaction_command.0;
+//         if command != "message" {
+//             continue;
+//         }
+
+//         let arg = &interaction_command.1;
+//         println!("{}", arg);
+//     }
+// }
```
I recommend removing this as well, since we'll be dealing with this in a future feature.

Exonoesis replied:
I agree, there's nothing we'd want to salvage from this function

divark commented on src/plugins/playable_character.rs:
```diff
@@ -8,28 +9,56 @@ pub struct PlayableCharacterPlugin;
+            OnEnter(AppState::InGame),
+            (load_player_movement_sound, load_player_bump_sound),
         )
+        //.add_systems(
+        //    Update,
+        //    (
+        //        move_entity,
+        //        animate_entity,
+        //        interact_entity,
+        //        display_interactive_message.after(interact_entity),
+        //        transition_level.after(interact_entity),
+        //        bound_player_movement,
+        //    )
+        //        .run_if(in_state(AppState::InGame)),
+        //)
```
This commented section can be removed.

Exonoesis replied:
Done

divark commented on src/plugins/playable_character.rs:
```diff
@@ -8,28 +9,56 @@ pub struct PlayableCharacterPlugin;
-        .add_event::<InteractionEvent>()
-        .register_ldtk_entity::<PlayerBundle>("Player");
+        .add_event::<PlayerMovementActions>();
+        //.add_event::<InteractionEvent>()
+        //.register_ldtk_entity::<PlayerBundle>("Player");
```
These comments can be removed as well.

Exonoesis replied:
And done

divark commented on src/plugins/playable_character.rs:
```diff
@@ -8,28 +9,56 @@ pub struct PlayableCharacterPlugin;
+            (set_player_target, move_entity_to_target).run_if(in_state(AppState::InGame)),
+        )
+        .add_event::<PlayerMovementActions>();
+        //.add_event::<InteractionEvent>()
```
This commented line can be removed as well.

Exonoesis replied:
And done again

divark commented on src/visuals/map.rs:
```diff
@@ -1,19 +1,1065 @@
+    interactive_markers
+}
+
+pub fn create_interactive_type(properties: &HashMap<String, PropertyValue>) -> InteractiveType {
+    let interactive_types: Vec<&String> = properties.keys().collect();
+    let interactive_values: Vec<&PropertyValue> = properties.values().collect();
+
+    // We assume that there is only one property on a marker
+    let interactive_type = interactive_types[0].clone();
+    let property_value = interactive_values[0].clone();
+    let mut property_string = "".to_string();
+
+    if let PropertyValue::StringValue(s) = property_value {
+        property_string = s;
+    }
+
+    match interactive_type.as_str() {
+        "Transition" => InteractiveType::Transition(PathBuf::from(property_string)),
+        _ => panic!("create_interactive_type: Marker Type is invalid"),
+    }
+}
```
```suggestion
pub fn create_interactive_type(properties: &HashMap<String, PropertyValue>) -> InteractiveType {
    let property_value = properties.get("Transition");

    // We assume that there is only one property on a marker
    if let Some(PropertyValue::StringValue(property_string)) = property_value {
        InteractiveType::Transition(PathBuf::from(property_string))
    } else {
        panic!("create_interactive_type: Marker Type is invalid")
    }
}
```

Exonoesis replied:
Oh, that's a lot cleaner, yeah

divark reviewed (changes requested):
Wow, what a rewrite! Both automatic and manual tests pass on my side. There are just a few areas that can be cleaned up.

Exonoesis committed:
Made PR suggested changes

- Removed LDtk map
- Removed various commented out code blocks
- Fixed typos
- Reworked create_interactive_type function for simplicity
